### DISCUSSION TITLE
CSS Color Variables: An easy way to set all the variations?

### DISCUSSION THREAD

**Discussion by ghost** [twbs/bootstrap#37349] [Q&A]
### TLDR
Is there some helper function(s) in JS, to set all of the color-variation CSS variables just by specifying the main colors, like `setBootstrapColorsCss(myPrimary, mySecondary, myWarning...)`?

### Long Version
I'm using Bootstrap 5.2.2 to display forms, but those forms need to have their colors set dynamically. For example, one customer's forms might have red as their primary color and orange as their warning color, but another customer might have olive green and neon green on their forms.

The problem is that Bootstrap currently uses SASS for all the color variations for hover, active, selected, border, background, for buttons, checks, radios, etc, but I need this to be at run-time with the CSS variables that Bootstrap uses. (According to [this](https://github.com/twbs/bootstrap/discussions/37268#discussioncomment-3816841) and [this](https://github.com/twbs/bootstrap/discussions/37057#discussioncomment-3519762), we might have global CSS variables like this in Bootstrap 6, but not yet.) The solution I'm using right now, is use jquery to set all the various color variations' CSS variables myself in javascript. This is pretty slow and error-prone, because I'm basically manually walking through all the SASS [rules](https://github.com/twbs/bootstrap/blob/2be9cd59bae8c864f6d2cfb03c535e96f00160ca/scss/_buttons.scss#L128-L131) and [mixins](https://github.com/twbs/bootstrap/blob/ac654a0f66e10dbe0db530d9058a6ba0d67d673a/scss/mixins/_buttons.scss#L21-L34), to get a list of all the color variables I need to override, to make myself code like:

```
export function setBootstrapColorCssVariables(form: HTMLElement, styles: Style) {
    const buttons = $(form).find('.btn-primary');

    buttons.css('--bs-btn-color', styles.onPrimaryColor);
    buttons.css('--bs-link-color', styles.onPrimaryColor);
    buttons.css('--bs-btn-bg', styles.primaryColor);
    buttons.css('--bs-btn-border-color', styles.primaryColor);

    // more here...

    const formControls = $(form).find('.form-control');
    formControls.focus(function () {
        const c = Color(styles.primaryColor);
        $(this).css("box-shadow", `0 0 0 .25rem rgba(${c.red()} ${c.green()} ${c.blue()} / 0.25)`);
        $(this).css("border-color", c.darken(0.5).hex());
    }).blur(function () {
        $(this).css("box-shadow", "")
        $(this).css("border-color", "");
    });
    
    // I'm still adding checkboxes and radios...
}
```

**Are there any existing JS helper-functions to set all of the CSS variables for these color variations, only needing to specify the main colors like `setBootstrapColorsCss(myPrimary, mySecondary, myWarning...)`?** I searched the discussions, StackOverflow for bootstrap 5 and the JS in this repo, but didn't find anything for setting CSS variables like this.

**Comment by mdo** [comment]
I think the best options will be from generating HSL colors and manipulating eh lightness level. I'm kicking myself for not doing this in v5 and choosing RGB, but alas there's always a new major version :). 

If you convert our default hex values to HSL, you can lighten and darken those relatively easily to generate the lighter and darker shades for hover/active/focus states. That's what I'm thinking and wishing for at least :).

**Reply by ghost** [reply]
That helps with the color values themselves, but I still need to walk through the SCSS looking for all the different variations that get used by active/hover/whatever X button/check/everything. (But also answers my question, insofar as there isn't a nice way to do this yet. Thanks! :)
